### DISCUSSION TITLE
MongoDB Upsert ignores primary key type when inserting

### DISCUSSION THREAD

**Discussion by ThatOneAwkwardGuy** [mikro-orm/mikro-orm#3646] [Q&A] [answered]
**Describe the bug**

1.  I have a base class called BaseSchemaString, it's "id" attribute is set to the following: 
```
  /**
   * The ID of the entity
   */
  @PrimaryKey({ name: '_id' })
  id: string = customAlphabet(alphanumeric, 21)();
```
2. All my Entities extend from this base schema, so when they are created their _id attribute is always a string and not an ObjectId
3. When upserting a schema that has no _id attribute, the _id generated is ObjectId and not string

I've had a look through the EntityGenerator but couldn't really figure much out on how it works.

**To Reproduce**
Steps to reproduce the behaviour:
I think the description above explains it well enough. 

**Expected behaviour**
I would expect the _id attribute generated to be of type string. I'm not sure how the EntityGenerator works, but I would expect it to, if provided with the Entity class, create a new instance of the Entity and merge it with the data provided to the upsert and then flush.

**Versions**

| Dependency | Version |
"typescript": "4.8.4",
"@mikro-orm/core": "5.5.1-dev.1",
"@mikro-orm/migrations": "^5.5.1-dev.1",
"@mikro-orm/mongo-highlighter": "1.0.0",
"@mikro-orm/mongodb": "5.5.1-dev.1",
"@mikro-orm/nestjs": "5.1.2",
"@mikro-orm/seeder": "^5.5.1-dev.1",

**Comment by B4nan** [comment]
Please provide reproduction as code, I do understand the problem, but you are probably doing something wrong in other places. The same use case is actually covered in the tests, so moving to discussions right ahead.

https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/upsert.test.ts

(note the PK being a `number`, not `ObjectId` - same entity is used for testing all the drivers)

> When upserting a schema that has no _id attribute, the _id generated is ObjectId and not string

Your schema always need to have `_id` attribute in mongo. Your entity definition above is also having one. This is even validated, you would get a metadata error if you don't define one.

> I'm not sure how the EntityGenerator works

FYI this has nothing to do with EntityGenerator - that is a tool for generating entity definitions from an existing schema, and it is currently SQL only feature.

**Comment by ThatOneAwkwardGuy** [comment, marked as answer]
Went through the test you provided and tried my best to break it, but it still worked. Then went through my own code and still didn't find anything. I'm using NestJS so I'm assuming that may be what's causing the problem. Regardless I realised the upsert didn't really work the way I wanted it to, I wanted it to work with compound indexes, so I ended up writing my own.

```
export const upsert = async <T extends object>(
  orm: MikroORM,
  entity: EntityClass<T>,
  data: any,
  uniqueKeys: (keyof T)[]
) => {
  const dataHasUniqueKeys = uniqueKeys.every((uniqueKey) =>
    has(data, uniqueKey)
  );
  if (dataHasUniqueKeys) {
    const result = await orm.em.findOne(
      entity.name,
      pick(data, uniqueKeys) as FilterQuery<T>
    );
    if (result) {
      return orm.em.assign(result, data, assignOptions);
    } else {
      return orm.em.create(entity, data);
    }
  }
  throw new Error('Data is missing unique keys');
};
```

Is it type-safe? anywhere from no to kind of. Will it do for now? Yea.
